src/gretel_synthetics/generator.py

```diff
@@ -0,0 +1,236 @@
+        return None
+
+
+class Generator:
+    """
+    A class for generating synthetic lines of text.
+
+    Text generation is initiated via the ``generate_next`` method, which returns an iterable that yields values
+    until the given number of _valid_ lines is returned. Each Generator also has a maximum budget of the number
+    of invalid lines that can be generated across _all_ calls to ``generate_next``.
+    """
+    settings: Settings
+    model: tf.keras.Sequential
+    sp: spm.SentencePieceProcessor
+    delim: str
+    total_invalid: int = 0
+    total_generated: int = 0
+
+    def __init__(self, settings: Settings):
+        """
+        Constructor.
+
+        Args:
+            settings: the generator settings to use.
+        """
+        self.settings = settings
+        self.sp, self.model = _load_model(settings.config)
+        self.delim = settings.config.field_delimiter
+
+    def generate_next(self, num_lines: int) -> Iterable[gen_text]:
+        """
+        Returns a sequence of lines.
+
+        Args:
+            num_lines: the number of _valid_ lines that should be generated during this call. The actual
+                number of lines returned may be higher, in case of invalid lines in the generation output.
+
+        Yields:
+            A ``gen_text`` object for every line (valid or invalid) that is generated.
+        """
+        lines_generated = 0
+        while lines_generated < num_lines:
+            rec = _predict_chars(self.model, self.sp, self.settings.start_string, self.settings.config).data
+            _valid = None
+            try:
+                if not self.settings.line_validator:
+                    yield gen_text(text=rec, valid=None, explain=None, delimiter=self.delim)
+                else:
+                    check = self.settings.line_validator(rec)
+                    if check is False:
+                        _valid = False
+                        self.total_invalid += 1
+                    else:
+                        _valid = True
+                    yield gen_text(text=rec, valid=_valid, explain=None, delimiter=self.delim)
+            except Exception as err:
+                # logging.warning(f'Line failed validation: {rec} errored with {str(err)}')
+                self.total_invalid += 1
+                yield gen_text(text=rec, valid=False, explain=str(err), delimiter=self.delim)
+            else:
+                if self.settings.line_validator and _valid:
+                    lines_generated += 1
+                elif not self.settings.line_validator:
+                    lines_generated += 1
+                else:
+                    ...
+
+            if self.total_invalid > self.settings.max_invalid:
+                raise RuntimeError("Maximum number of invalid lines reached!")
+
+
+def _predict_chars(
+    model: tf.keras.Sequential,
+    sp: spm.SentencePieceProcessor,
+    start_string: str,
+    store: "BaseConfig",
```

Comment:
use obj reference here

Reply:
I used the `if TYPE_CHECKING:` pattern, which I saw in an existing file - this makes it necessary to put type names in quotes (see https://mypy.readthedocs.io/en/stable/common_issues.html#import-cycles). Even though we do not need this to solve the issue of cyclical deps, I was assuming it was preferable to guard an `import` statement with this `if` clause if it's needed exclusively for type annotations. I can change it if that isn't the case.

Reply:
if you set the `BaseConfig` to `None` during run time (i.e. in an else statement from the `if TYPE CHECKING:`) it doesn't hurt anything at run time but gives the better dev UX.

Reply:
Sure, done now.